src/VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs

```diff
@@ -17,99 +20,108 @@ public class PersonalizationService : ITaggedItemService, ITaggedItemSearchServi
     {
         private readonly Func<IPersonalizationRepository> _repositoryFactory;
         private readonly ITaggedEntitiesServiceFactory _taggedEntitiesServiceFactory;
+        private readonly IPlatformMemoryCache _platformMemoryCache;
         private readonly ITagPropagationPolicy _tagPropagationPolicy;
 
         public PersonalizationService(Func<IPersonalizationRepository> repositoryFactory,
             ITagPropagationPolicy tagPropagationPolicy,
-            ITaggedEntitiesServiceFactory taggedEntitiesServiceFactory)
+            ITaggedEntitiesServiceFactory taggedEntitiesServiceFactory,
+            IPlatformMemoryCache platformMemoryCache)
         {
             _repositoryFactory = repositoryFactory;
             _tagPropagationPolicy = tagPropagationPolicy;
             _taggedEntitiesServiceFactory = taggedEntitiesServiceFactory;
+            _platformMemoryCache = platformMemoryCache;
         }
 
         public async Task<GenericSearchResult<TaggedItem>> SearchTaggedItemsAsync(TaggedItemSearchCriteria criteria)
         {
-            var result = new GenericSearchResult<TaggedItem>();
 
+            var cacheKey = CacheKey.With(GetType(), nameof(SearchTaggedItemsAsync), criteria.GetCacheKey());
+            return await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey,
+                async (cacheEntry) =>
+                {
+                    cacheEntry.AddExpirationToken(PersonalizationCacheRegion.CreateChangeToken());
```

Comment:
Need to eliminate `GenericSearchResult<TaggedItem>` in favor of `TaggedItemSearchResult`.
Controller uses it.